src/metric-engine/src/data_region.rs

```diff
@@ -65,33 +61,30 @@ impl DataRegion {
     pub async fn add_columns(
         &self,
         region_id: RegionId,
-        columns: &mut [ColumnMetadata],
+        columns: Vec<ColumnMetadata>,
         index_options: IndexOptions,
     ) -> Result<()> {
+        // Return early if no new columns are added.
+        if columns.is_empty() {
+            return Ok(());
+        }
+
         let region_id = utils::to_data_region_id(region_id);
 
-        let mut retries = 0;
-        // submit alter request
-        while retries < MAX_RETRIES {
-            let request = self
-                .assemble_alter_request(region_id, columns, index_options)
-                .await?;
+        let num_columns = columns.len();
+        let request = self
+            .assemble_alter_request(region_id, columns, index_options)
+            .await?;
 
-            let _timer = MITO_DDL_DURATION.start_timer();
+        let _timer = MITO_DDL_DURATION.start_timer();
 
-            let result = self.mito.handle_request(region_id, request).await;
-            match result {
-                Ok(_) => return Ok(()),
-                Err(e) if e.status_code() == StatusCode::RequestOutdated => {
-                    info!("Retrying alter {region_id} due to outdated schema version, times {retries}");
-                    retries += 1;
-                    continue;
-                }
-                Err(e) => {
-                    return Err(e).context(MitoWriteOperationSnafu)?;
-                }
-            }
-        }
+        let _ = self
+            .mito
+            .handle_request(region_id, request)
+            .await
```

Comment:
Question: why are we removing the retry logic?

Reply:
There are two main reasons. First, retries might be incorrect if the schema version changes, as we need to filter out potential duplicate columns. Second, procedure locks ensure that only one exclusive operation is allowed at a time.